### PR TITLE
drivers/exec+java: Add configuration to restore previous PID/IPC namespace behavior

### LINKED ISSUE #9969: Configuration for task namespace isolation in driver plugin
Nomad v1.0.3 fixes a security issue where task PID and IPC namespaces were shared among tasks using the same user. To re-enable the previous behavior, we can add configuration options to the exec-based driver plugins, e.g. 

`default_pid_mode`: "private" or "host" [default: "private"]
`default_ipc_mode`: "private" or "host" [default: "private"]

These would be applied cluster-wide to all tasks using an exec-based task driver (e.g. `exec` and `java`).

### PR DESCRIPTION
**Pull request by shoenig**
This PR adds default_pid_mode and default_ipc_mode options to the exec and java
task drivers. By default these will default to "private" mode, enabling PID and
IPC isolation for tasks. Setting them to "host" mode disables isolation. Doing
so is not recommended, but may be necessary to support legacy job configurations.

Closes #9969

### CODE REVIEW

**Comment by shoenig** [comment]
## spot check exec

`exec.nomad`

```hcl
job "exec" {
  datacenters = ["dc1"]

  type = "batch"

  group "catter" {
    task "cat" {
      driver = "exec"
      config {
	command = "/usr/bin/cat"
	args = ["/proc/self/stat"]
      }
    }
  }
}
```

#### private mode

`nomad.hcl`

```hcl
plugin "exec" {
  config {
    default_pid_mode = "private"
    default_ipc_mode = "private"
  }
}
```

`/proc/self/stat` indicates PID 1, i.e. in a namespace

```
[x52 nsiso-driver (master)] $ nomad job run exec.nomad
==> Monitoring evaluation "bc3a4e30"
    Evaluation triggered by job "exec"
    Allocation "7963d179" created: node "02912ff7", group "catter"
==> Monitoring evaluation "bc3a4e30"
    Evaluation status changed: "pending" -> "complete"
==> Evaluation "bc3a4e30" finished with status "complete"
[x52 nsiso-driver (master)] $ nomad alloc logs 796
1 (cat) R 0 1 1 0 -1 4194560 1701 0 0 0 0 1 0 0 20 0 1 0 16268147 8425472 146 18446744073709551615 94617221357568 94617221375025 140735364814288 0 0 0 0 0 0 0 0 0 17 3 0 0 0 0 0 94617221393040 94617221394624 94617247051776 140735364822617 140735364822646 140735364822646 140735364825067 0
```

#### host mode

`nomad.hcl`

```hcl
plugin "exec" {
  config {
    default_pid_mode = "host"
    default_ipc_mode = "host"
  }
}
```

`/proc/self/stat` indicates PID 214264, i.e. sharing host PID namespace

```
[x52 nsiso-driver (master)] $ nomad job run exec.nomad
==> Monitoring evaluation "75d89109"
    Evaluation triggered by job "exec"
    Allocation "07d5cc15" created: node "063de9b8", group "catter"
==> Monitoring evaluation "75d89109"
    Evaluation status changed: "pending" -> "complete"
==> Evaluation "75d89109" finished with status "complete"
[x52 nsiso-driver (master)] $ nomad alloc logs 07d
214264 (cat) R 214252 214264 214264 0 -1 4194560 1670 0 0 0 1 0 0 0 20 0 1 0 16277382 8425472 130 18446744073709551615 93974594875392 93974594892849 140729171405200 0 0 0 0 0 0 0 0 0 17 3 0 0 0 0 0 93974594910864 93974594912448 93974621917184 140729171408473 140729171408502 140729171408502 140729171410923 0
```

**Comment by shoenig** [comment]
## spot check java

`PID.java`

```java
import java.io.File;
import java.io.FileInputStream;

public class PID {
    public static void main(String[] args) throws Exception {
	int pid = Integer.parseInt(new File("/proc/self").getCanonicalFile().getName());
	System.out.println(pid);
    }
}
```

`java.nomad`

```hcl
job "java" {
  datacenters = ["dc1"]

  type = "batch"

  group "printer" {
    task "print" {
      driver = "java"
      config {
	class = "PID"
	class_path = "local/"
      }

      artifact {
	source = "http://localhost:8000/PID.class"
      }
    }
  }
}
```

compile & serve class file

```
[x52 nsiso-driver (master)] $ javac PID.java
[x52 nsiso-driver (master)] $ python3 -m http.server 8000
Serving HTTP on 0.0.0.0 port 8000 (http://0.0.0.0:8000/) ...
```

### private mode

`nomad.hcl`

```hcl
plugin "java" {                                                                                                                                                               
  config {                                                                                                                                                                    
    default_pid_mode = "private"                                                                                                                                              
    default_ipc_mode = "private"                                                                                                                                              
  }                                                                                                                                                                           
} 
```

PID 1 indicates isolated namespace

```
[x52 nsiso-driver (master)] $ nomad job run java.nomad
==> Monitoring evaluation "3e0fa2b3"
    Evaluation triggered by job "java"
    Allocation "d9cc2e91" created: node "25facb9e", group "printer"
==> Monitoring evaluation "3e0fa2b3"
    Evaluation status changed: "pending" -> "complete"
==> Evaluation "3e0fa2b3" finished with status "complete"
[x52 nsiso-driver (master)] $ nomad alloc logs d9c
1
```

### host mode

`nomad.hcl`

```hcl
plugin "java" {
  config {
    default_pid_mode = "host"
    default_ipc_mode = "host"
  }
}

```

PID of 215856 indicates using host PID namespace

```
[x52 nsiso-driver (master)] $ nomad job run java.nomad
==> Monitoring evaluation "dd73c70a"
    Evaluation triggered by job "java"
    Allocation "722372e3" created: node "aa4179ca", group "printer"
==> Monitoring evaluation "dd73c70a"
    Evaluation status changed: "pending" -> "complete"
==> Evaluation "dd73c70a" finished with status "complete"
[x52 nsiso-driver (master)] $ nomad alloc logs 722
215856
```

File: drivers/exec/driver.go [hunk 1]
```diff
@@ -122,6 +130,30 @@ type Config struct {
 	// NoPivotRoot disables the use of pivot_root, useful when the root partition
 	// is on ramdisk
 	NoPivotRoot bool `codec:"no_pivot_root"`
+
+	// DefaultModePID is the default PID isolation set for all tasks using
+	// exec-based task drivers.
+	DefaultModePID string `codec:"default_pid_mode"`
```
Comment (krishicks) [inline drivers/exec/driver.go:136]: What do you think about using `DefaultPIDMode` here, to keep parity with the config key `default_pid_mode`?

File: drivers/shared/executor/executor_linux.go [hunk 2]
```diff
@@ -578,12 +589,8 @@ func configureIsolation(cfg *lconfigs.Config, command *ExecCommand) error {
 	// disable pivot_root if set in the driver's configuration
 	cfg.NoPivotRoot = command.NoPivotRoot
 
-	// launch with mount namespace
-	cfg.Namespaces = lconfigs.Namespaces{
-		{Type: lconfigs.NEWNS},
-		{Type: lconfigs.NEWPID},
-		{Type: lconfigs.NEWIPC},
-	}
+	// setup default namespaces as configured
```
Comment (krishicks) [inline drivers/shared/executor/executor_linux.go:592]:
```suggestion
	// set up default namespaces as configured
```

File: drivers/shared/executor/executor.go [hunk 3]
```diff
@@ -34,6 +34,12 @@ const (
 	// ExecutorVersionPre0_9 is the version of executor use prior to the release
 	// of 0.9.x
 	ExecutorVersionPre0_9 = "1.1.0"
+
+	// IsoModePrivate represents the private isolation mode for a namespace
+	IsoModePrivate = "private"
```
Comment (krishicks) [inline drivers/shared/executor/executor.go:39]: I don't feel like "Iso" is a common short form of Isolation/Isolate in the codebase; would it be possible to expand this to `IsolationModePrivate`?

File: drivers/shared/executor/executor.go [hunk 4]
```diff
@@ -132,7 +138,14 @@ type ExecCommand struct {
 	// Devices are the the device nodes to be created in isolation environment
 	Devices []*drivers.DeviceConfig
 
+	// NetworkIsolation is the network isolation configuration.
 	NetworkIsolation *drivers.NetworkIsolationSpec
+
+	// DefaultModePID is the default PID isolation mode (private or host).
```
Comment (krishicks) [inline drivers/shared/executor/executor.go:144]: I feel like the "(private or host)" part of this comment should be removed; if there was a future mode other than private or host, this comment would become out of date, but nobody would know.

Also, comments aren't meant to have full stops at the end.

**Review by krishicks — APPROVED** [review]

File: drivers/exec/driver_test.go [hunk 5]
```diff
@@ -273,7 +275,7 @@ func TestExecDriver_StartWaitRecover(t *testing.T) {
 // task dies, the orphans in the PID namespaces are killed by the kernel
 func TestExecDriver_NoOrphans(t *testing.T) {
 	t.Parallel()
-	require := require.New(t)
+	r := require.New(t)
```
Comment (cgbaker) [inline drivers/exec/driver_test.go:278]: are we doing this now?

File: website/content/docs/drivers/exec.mdx [hunk 6]
```diff
@@ -107,7 +107,21 @@ This also applies for running Nomad in -dev mode.
 
 ## Plugin Options
 
-- `no_pivot_root` - Defaults to `false`. When `true`, the driver uses `chroot`
+- `default_pid_mode` `(string: optional)` - Defaults to `"private"`. Set to
+ `"private"` to enable namespace isolation for tasks by default, or `"host"` to
```
Comment (cgbaker) [inline website/content/docs/drivers/exec.mdx:111]:
```suggestion
 `"private"` to enable PID namespace isolation for tasks by default, or `"host"` to
```

File: website/content/docs/drivers/exec.mdx [hunk 7]
```diff
@@ -107,7 +107,21 @@ This also applies for running Nomad in -dev mode.
...
+ disable isolation.
+
+!> **Warning:** If set to `"host"`, other processes running as the same user will
+ be able to access sensitive process information like environment variables.
+
+- `default_ipc_mode` `(string: optional)` - Defaults to `"private"`. Set to
+`"private"` to enable inter-process-communication isolation for tasks by default,
```
Comment (cgbaker) [inline website/content/docs/drivers/exec.mdx:118]:
```suggestion
`"private"` to enable IPC namespace isolation for tasks by default,
```

File: website/content/docs/drivers/java.mdx [hunk 8]
```diff
@@ -109,6 +109,22 @@ The `java` driver implements the following [capabilities](/docs/internals/plugin
 | network isolation    | host, group                   |
 | volume mounting      | none, all (only for linux)    |
 
+## Plugin Options
+
+- `default_pid_mode` `(string: optional)` - Defaults to `"private"`. Set to
+ `"private"` to enable namespace isolation for tasks by default, or `"host"` to
```
Comment (cgbaker) [inline website/content/docs/drivers/java.mdx:115]:
```suggestion
 `"private"` to enable PID namespace isolation for tasks by default, or `"host"` to
```

File: website/content/docs/drivers/java.mdx [hunk 9]
```diff
@@ -109,6 +109,22 @@ The `java` driver implements the following [capabilities](/docs/internals/plugin
...
+ disable isolation.
+
+!> **Warning:** If set to `"host"`, other processes running as the same user will
+ be able to access sensitive process information like environment variables.
+
+- `default_ipc_mode` `(string: optional)` - Defaults to `"private"`. Set to
+`"private"` to enable inter-process-communication isolation for tasks by default,
```
Comment (cgbaker) [inline website/content/docs/drivers/java.mdx:122]:
```suggestion
`"private"` to enable IPC namespace isolation for tasks by default,
```

File: website/content/docs/drivers/java.mdx [hunk 10]
```diff
@@ -109,6 +109,22 @@ The `java` driver implements the following [capabilities](/docs/internals/plugin
...
+ or `"host"` to disable isolation.
+
+!> **Warning:** If set to `"host"`, other processes running as the same user will be
+ able to make use of IPC features, like sending unexpected posix signals.
```
Comment (cgbaker) [inline website/content/docs/drivers/java.mdx:126]:
```suggestion
 able to make use of IPC features, like sending unexpected POSIX signals.
```

File: website/content/docs/drivers/exec.mdx [hunk 11]
```diff
@@ -107,7 +107,21 @@ This also applies for running Nomad in -dev mode.
...
+ or `"host"` to disable isolation.
+
+!> **Warning:** If set to `"host"`, other processes running as the same user will be
+ able to make use of IPC features, like sending unexpected posix signals.
```
Comment (cgbaker) [inline website/content/docs/drivers/exec.mdx:122]:
```suggestion
 able to make use of IPC features, like sending unexpected POSIX signals.
```

**Review by cgbaker — APPROVED** [review]
LGTM, thanks, @shoenig

File: drivers/exec/driver_test.go (hunk 5 above)
Comment (shoenig) [reply drivers/exec/driver_test.go:278]: I only care that we do not shadow names. Whether we use `require.Blah(t, ...` or rename `require` as a variable or something else, I don't have much of an opinion.

**Comment by krishicks** [comment]
p.s. I really like the spot checks. So nice to see that in a PR!

**Pull request merged**